fix: canonicalization of chunked ExtensionArray

a10y opened the pull request:
Since #346, we've been canonicalizing chunked ExtensionArray's incorrectly. Rather than unwrapping each chunk to its backing storage array we've just been building a new chunkedarray of the ExtensionArray chunks.

Before #480 , this was actually causing DateTimePartsCompressor to fail in can_compress, so we weren't going down the compression codepath.

## The Fix

Fix `into_canonical` so that when we encounter a `ChunkedArray(ExtensionArray(storage))` to yield an `ExtensionArray(ChunkedArray(storage))` where each chunk is canonicalized first (e.g. if you have chunks of DateTimePartsArray you will end up with chunks of ExtensionArray(i64)).

Fix the DateTimePartsCompressor to canonicalize its input, to handle that case where it may be a ChunkedArray.

a10y committed:
save

a10y commented on encodings/datetime-parts/src/compress.rs:
```diff
@@ -2,15 +2,19 @@ use vortex::array::datetime::{TemporalArray, TimeUnit};
-    let timestamps = try_cast(&array.temporal_values(), PType::I64.into())?.into_primitive()?;
+    let timestamps = try_cast(
+        &array.temporal_values(),
+        &DType::Primitive(PType::I64, Nullability::Nullable),
```
`CastFn` has poorly defined semantics for what the DType of the resulting array is, which is causing panic when building new chunkedarray. We should probably choose a semantic and document it as part of #260

a10y replied:
this is no longer a problem now that we're not pushing chunking internally, so i'm removing this

robert3005 commented on vortex-array/src/array/chunked/canonical.rs:
```diff
@@ -47,9 +49,14 @@ pub(crate) fn try_canonicalize_chunks(
         // it is safe to unpack them.
         DType::Extension(ext_dtype, _) => {
+            let storage_chunks: Vec<Array> = chunks
+                .iter()
+                .map(|chunk| ExtensionArray::try_from(chunk).unwrap().storage())
+                .collect();
+            let storage_dtype = storage_chunks.first().unwrap().dtype().clone();
             let ext_array = ExtensionArray::new(
                 ext_dtype.clone(),
-                ChunkedArray::try_new(chunks, dtype.clone())?.into_array(),
+                ChunkedArray::try_new(storage_chunks, storage_dtype)?.into_array(),
```
hmm, maybe this is enough. Not sure though since it behaves differently than any other canonical type

gatesn commented on encodings/datetime-parts/src/array.rs:
```diff
@@ -11,7 +11,7 @@ use vortex_error::{vortex_bail, VortexResult};
 use crate::compute::decode_to_temporal;
 
-impl_encoding!("vortex.datetimeparts", 22u16, DateTimeParts);
+impl_encoding!("vortex.datetimeparts", 23u16, DateTimeParts);
```
?

gatesn replied:
Can we create a FLUP to make this an enum

robert3005 replied:
This cannot be an enum…

robert3005 replied:
Rather this shouldn’t be an enum. You end up with the same problem and false sense of security

gatesn replied:
Oh sorry, I don't mean an enum, I just mean centralized constants

a10y committed:
canonicalize before ExtensionArray::try_from

a10y committed:
fix

robert3005 commented on encodings/datetime-parts/src/array.rs:
```diff
@@ -11,7 +11,7 @@ use vortex_error::{vortex_bail, VortexResult};
 use crate::compute::decode_to_temporal;
 
-impl_encoding!("vortex.datetimeparts", 22u16, DateTimeParts);
+impl_encoding!("vortex.datetimeparts", 23u16, DateTimeParts);
```
remove this

a10y committed:
canonicalize Extension-typed things recursively

a10y committed:
revert this one file to develop

a10y committed:
don't need this anymore either

a10y committed:
revert a bunch of stuff

a10y commented on vortex-array/src/array/chunked/canonical.rs:
```diff
@@ -28,6 +28,8 @@ pub(crate) fn try_canonicalize_chunks(
     chunks: Vec<Array>,
     dtype: &DType,
 ) -> VortexResult<Canonical> {
+    assert!(!chunks.is_empty(), "chunks must be non-empty");
```
i do this b/c i access storage_chunks.first() below

lwwmanning commented on vortex-array/src/array/chunked/canonical.rs:
```diff
@@ -28,6 +28,8 @@ pub(crate) fn try_canonicalize_chunks(
     chunks: Vec<Array>,
     dtype: &DType,
 ) -> VortexResult<Canonical> {
+    assert!(!chunks.is_empty(), "chunks must be non-empty");
```
is it crash-worthy programmer error if it's empty? if not, should return an `Err` instead

a10y replied:
done

a10y committed:
post bail

gatesn commented on vortex-array/src/array/chunked/canonical.rs:
```diff
@@ -44,15 +46,26 @@ pub(crate) fn try_canonicalize_chunks(
+                .iter()
+                // Extension-typed arrays can be compressed into something that is not an
+                // ExtensionArray, so we should canonicalize each chunk into ExtensionArray first.
+                .map(|chunk| chunk.clone().into_extension().unwrap().storage())
```
My preference would be to remove the into_extension functions, since they're confusing alongside the variant traits. `ExtensionArray::try_from(chunk.clone())`

a10y replied:
this is actually different from just doing ExtensionArray::try_from. `.into_extension()` actually calls `.into_canonical()` first

gatesn replied:
Ohhh gotcha. Canonicalizing makes sense. But in that case, I dislike the function name even more.... We should just be explicit and not have these shorthands (but that's a change for another day)

a10y replied:
yea I also find them footgun-y and would prefer to remove them, I think @robert3005 was partial to them though

gatesn replied:
Maybe `into_canonical_extension()` is better, or `canonicalize_extension()`

robert3005 replied:
We probably should rename them but the previous version of two function calls was ugly imho

gatesn replied:
Maybe, but explicit is better than ugly

a10y replied:
i filed #507 to track a flup on this

gatesn commented on vortex-array/src/array/chunked/canonical.rs:
```diff
@@ -28,6 +28,10 @@ pub(crate) fn try_canonicalize_chunks(
     chunks: Vec<Array>,
     dtype: &DType,
 ) -> VortexResult<Canonical> {
+    if chunks.is_empty() {
```
You should add a FLUP to create `Canonical::empty(dtype)`

a10y replied:
If we end up redesigning extension arrays to contain their storage_dtype as part of the type definition, i'm not sure we'd need it

gatesn commented on vortex-array/src/array/chunked/canonical.rs:
```diff
@@ -44,15 +48,26 @@ pub(crate) fn try_canonicalize_chunks(
+                // ExtensionArray, so we should canonicalize each chunk into ExtensionArray first.
+                .map(|chunk| chunk.clone().into_extension().unwrap().storage())
+                .collect();
+            let storage_dtype = storage_chunks.first().unwrap().dtype().clone();
```
This is weird...

What if we ask the extension array for its canonical DType? I guess we could still get a different answer for every chunk.

Is the constraint that all chunks of a chunked array must have the same dtype?

a10y replied:
the storage DType for an extensionarray currently is something that is only known at runtime, which is why we need to do this.

ChunkedArray's constructor enforces that all chunks must have same DType, but again the storage_dtype is held in metadata, not on DType, so there is nothing protecting against that

a10y committed:
leave canonical ExtensionArray->storage as ChunkedArray

a10y committed:
fix comment

a10y commented:
Per discussion on Slack yesterday, I updated the branch to have canonicalization behavior Chunked(Extension(storage)) -> Extension(Chunked(storage)). As part of that, I added call to .into_canonical() for the DateTimePartsCompressor.

robert3005 commented:
Should change the commit message since the behaviour is different than what we started with
